library-model/library-model-generator/src/main/java/com/uber/nullaway/libmodel/LibraryModelGenerator.java

```diff
@@ -0,0 +1,274 @@
+    Path outputPathInstance = Paths.get(outputPath);
+    try {
+      Files.createDirectories(outputPathInstance.getParent());
+      try (DataOutputStream dos = new DataOutputStream(Files.newOutputStream(outputPathInstance))) {
+        StubxWriter.write(
+            dos,
+            importedAnnotations,
+            Collections.emptyMap(),
+            Collections.emptyMap(),
+            methodRecords);
+      }
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  public Path dirnameToPath(String dir) {
+    File f = new File(dir);
+    String absoluteDir = f.getAbsolutePath();
+    if (absoluteDir.endsWith("/.")) {
+      absoluteDir = absoluteDir.substring(0, absoluteDir.length() - 2);
+    }
+    return Paths.get(absoluteDir);
+  }
+
+  private static class AnnotationCollectorCallback implements SourceRoot.Callback {
+
+    private final AnnotationCollectionVisitor annotationCollectionVisitor;
+
+    public AnnotationCollectorCallback(Map<String, MethodAnnotationsRecord> methodRecords) {
+      this.annotationCollectionVisitor = new AnnotationCollectionVisitor(methodRecords);
+    }
+
+    @Override
+    public Result process(Path localPath, Path absolutePath, ParseResult<CompilationUnit> result) {
+      Result res = Result.SAVE;
+      Optional<CompilationUnit> opt = result.getResult();
+      if (opt.isPresent()) {
+        CompilationUnit cu = opt.get();
+        cu.accept(annotationCollectionVisitor, null);
+      }
+      return res;
+    }
+  }
+
+  private static class AnnotationCollectionVisitor extends VoidVisitorAdapter<Void> {
+
+    private String packageName = "";
+    private boolean isJspecifyNullableImportPresent = false;
+    private boolean isNullMarked = false;
+    private Map<String, MethodAnnotationsRecord> methodRecords;
+    private static final String ARRAY_RETURN_TYPE_STRING = "Array";
+    private static final String NULL_MARKED = "NullMarked";
+    private static final String NULLABLE = "Nullable";
+    private static final String JSPECIFY_NULLABLE_IMPORT = "org.jspecify.annotations.Nullable";
+
+    public AnnotationCollectionVisitor(Map<String, MethodAnnotationsRecord> methodRecords) {
+      this.methodRecords = methodRecords;
+    }
+
+    @Override
+    public void visit(PackageDeclaration pd, Void arg) {
+      this.packageName = pd.getNameAsString();
+      super.visit(pd, null);
+    }
+
+    @Override
+    public void visit(ImportDeclaration id, Void arg) {
+      if (id.getName().toString().contains(JSPECIFY_NULLABLE_IMPORT)) {
+        this.isJspecifyNullableImportPresent = true;
+      }
+      super.visit(id, null);
+    }
+
+    @Override
+    public void visit(ClassOrInterfaceDeclaration cid, Void arg) {
+      cid.getAnnotations()
+          .forEach(
+              a -> {
+                if (a.getNameAsString().equalsIgnoreCase(NULL_MARKED)) {
+                  this.isNullMarked = true;
+                }
```

Comment:
This logic won't handle cases where `@NullMarked` appears on some nested classes but not others.  But, I think for now, we can just say we only support `@NullMarked` on the top-level class.  Can we add a comment to this effect?

Reply:
It also needs `@NullMarked` on the same file, correct? We don't have a way to add this to the package somewhere and have the rest of the traversal be aware of that (that seems more likely than mixed marked/unmarked code in the models, but I am not sure)

Reply:
Yes, you're correct, @lazaroclapp, we are assuming an explicit `@NullMarked` on the top-level class in a source file, and that there is only one top-level class per source file.  We won't find anything on `package-info.java` or `module-info.java` for now.  Both these assumptions hold for jspecify/jdk

Reply:
@akulk022 can you update the comment in light of the above?